plugin/core/diagnostics_manager.py

```diff
@@ -17,44 +20,83 @@ class DiagnosticsManager(OrderedDict):
+        diagnostic for this `uri` with `pred(diagnostic) == True`, filtered by `bool(f(diagnostic))`.
+        Only `uri`s with non-empty `results` are returned. Each `uri` is guaranteed to be yielded
+        not more than once. Items and results are ordered as they came in from the server.
+        """
+        for uri, diagnostics in self.items():
+            results = list(filter(None, map(functools.partial(f, uri), filter(pred, diagnostics))))  # type: List[T]
+            if results:
+                yield uri, results
 
-    def diagnostics_panel_contributions_async(
-        self,
-    ) -> Iterator[Tuple[str, List[Tuple[str, Optional[int], Optional[str], Optional[str]]]]]:
-        for path, (contribution, _, _) in self.items():
-            if contribution:
-                yield path, contribution
+    def filter_map_diagnostics_flat_async(self, pred: Callable[[Diagnostic], bool],
+                                          f: Callable[[ParsedUri, Diagnostic], T]) -> Iterator[Tuple[ParsedUri, T]]:
+        """
+        Flattened variant of `filter_map_diagnostics_async()`. Yields `(uri, result)` items for each
+        of the `result`s per `uri` instead. Each `uri` can be yielded more than once. Items are
+        grouped by `uri` and each `uri` group is guaranteed to appear not more than once. Items are
+        ordered as they came in from the server.
+        """
+        for uri, results in self.filter_map_diagnostics_async(pred, f):
+            for result in results:
+                yield uri, result
 
     def sum_total_errors_and_warnings_async(self) -> Tuple[int, int]:
+        """
+        Returns `(total_errors, total_warnings)` count of all diagnostics currently in store.
+        """
         return (
-            sum(errors for _, errors, _ in self.values()),
-            sum(warnings for _, _, warnings in self.values()),
+            sum(map(severity_count(DiagnosticSeverity.Error), self.values())),
+            sum(map(severity_count(DiagnosticSeverity.Warning), self.values())),
         )
 
+    def diagnostics_by_document_uri(self, document_uri: DocumentUri) -> List[Diagnostic]:
+        """
+        Returns possibly empty list of diagnostic for `document_uri`.
+        """
+        return self.get(parse_uri(document_uri), [])
+
+    def diagnostics_by_parsed_uri(self, uri: ParsedUri) -> List[Diagnostic]:
+        """
+        Returns possibly empty list of diagnostic for `uri`.
+        """
+        return self.get(uri, [])
+
+
+def severity_count(severity: int) -> Callable[[List[Diagnostic]], int]:
+    def severity_count(diagnostics: List[Diagnostic]) -> int:
+        return len(list(filter(has_severity(severity), diagnostics)))
+
+    return severity_count
+
 
 def has_severity(severity: int) -> Callable[[Diagnostic], bool]:
     def has_severity(diagnostic: Diagnostic) -> bool:
         return diagnostic_severity(diagnostic) == severity
+
     return has_severity
```

Comment:
Maybe turn this into a lambda as well?

```suggestion
def has_severity(severity: int) -> Callable[[Diagnostic], bool]:
    return lambda diagnostic: diagnostic_severity(diagnostic) == severity
```

Reply:
These I would leave as they are because it's immediately clear from the layout that they are just curried functions, with lambdas this would be a little harder to get.

A proper `@curry` decorator would be nice, though:
```python
@curry
def is_severity_included(max_severity: ..., diagnostic: ...) -> ...:
    return diagnostic_severity(diagnostic) <= max_severity
```
There is one in the `toolz` package, but it seems overkill here and I don't know how well it plays with `typing`.

Reply:
I'm personally not accustomed to the use of curried functions. I see a callable as return type and would expect a `lambda` if it's a single line. I think this is true for a lot of Python (or imperative) developers. But, I guess your argumentation makes sense as well.

Reply:
One downside of lambdas is that those can't be type-annotated properly.

Reply:
I would recommend `functools.partial`, but iirc mypy also had problems with that. No idea about pyright.

Currying is really cool, but it's not native to Python and will forever feel alien unless it becomes integrated and more Python developers get to know it.